File: Keppnisforritun/vika5/knapsack.py

def knap(capacity, length, weights, values):
    # setja upp tvívítt fylki með indexi og þyngdum
    for index in range(length+1):
        for thyngd in range(capacity+1):
            #print("index og thyngd:", index, thyngd)
            #print("current geymsla:", geymsla)
            if index == 0 or thyngd == 0:
                geymsla[index][thyngd] = 0
            elif weights[index-1] <= thyngd:
                #print("er inni elif")
                #print("max:", max(values[index-1] + geymsla[index-1][thyngd - weights[index-1]],
                #geymsla[index-1][thyngd]))

                geymsla[index][thyngd] = max(values[index-1] + geymsla[index-1][thyngd - weights[index-1]],
                geymsla[index-1][thyngd])
            else:
                #print("er inni else")
                geymsla[index][thyngd] = geymsla[index-1][thyngd]
    
    #print("geymsla:",geymsla)
    # fara í gegnum fylkið afturábak og finna indexa sem passa best
    thyngd = capacity
    curr = geymsla[length][capacity]
    count = 0
    indices = []
    #print("curr:", geymsla[n][capacity])
    for i in range(length, 0, -1):
        #print(geymsla[i][thyngd])
        #print(geymsla)
        # if curr <= 0:
        #     break
            #print("er i curr <= 0 og i:", i)
        #print("i-1 og thyngd:", (i-1), thyngd)
        if curr == geymsla[i-1][thyngd]:
            #print("er i geymsla[i-1][thyngd] == curr")
            continue
        else:
            #print(weights[i-1])
            thyngd = thyngd - weights[i-1]
            curr = curr - values[i-1]
            count += 1
            indices.append((i-1))
    print(count)
    strengur = ""
    for x in range(len(indices)):
        print(indices[x], end=" ")
  
    
            
while True:
    data = []
    data = input().split()
    capacity = int(data[0])
    n = int(data[1])
    weights = []
    values = []
    geymsla = []
    indices = []
    count = 0
    for x in range(n):
        data = input().split()
        values.append(int(data[0]))
        weights.append(int(data[1]))
    #geymsla = [[0]*(capacity+1)]*(n+1)
    geymsla = [[0 for x in range((capacity + 1))] for y in range((n + 1))]
    #print(n)
    #print(geymsla)
    #print(weights)
    #print(values)
   
    knap(capacity, n, weights, values)

    #print(geymsla)


# # Brute !
# def knapBrute(capacity, i, weights, values):
#     if capacity == 0 or i == 0:
#         return 0
#     if capacity < weights[i-1]:
#         return knapBrute(capacity, i-1, weights, values)
#     else:
#         return max(knapBrute(capacity, i-1, weights, values),
#                 knapBrute(capacity - weights[i], i-1, weights, values) + values[i])
# for test in range(30):
#     data = input().split()
#     capacity = int(data[0])
#     n = int(data[1])
#     weights = []
#     values = [] = []
#     for x in range(n):
#         data = input().split()
#         values.append(int(data[0]))
#         weights.append(int(data[1]))
#     print(knapBrute(capacity, n-1, weights, values))


    

File: Keppnisforritun/vika5/knapsack.cpp
#include <iostream>
#include <vector>
#include <utility>
#include <stdio.h> 
#include <iostream>
using namespace std;




int knap(int capacity, int length,  vector<int> weights, vector<int> values){
    vector<vector<int>> storage(length+1, vector<int>(capacity+1));
    for(int i=0; i < length+1; i++){
        for(int thyngd=0; thyngd < capacity; thyngd++){
            if(i == 0 || thyngd == 0){
                storage[i][thyngd] = 0;
            }
            else if(weights[i-1] <= thyngd){
                storage[i][thyngd] = max(storage[i-1][thyngd], values[i-1] + storage[i-1][thyngd - weights[i-1]]
                );
            }
            else{
                storage[i][thyngd = storage[i-1][thyngd]];
            }
        }
    }
    
    int thyngd = capacity;
    int curr = storage[length][capacity];
    int count = 0;
    vector<int> indices;

    for(int i=length; i>0 && curr >= 1; i--){
        if (curr == storage[i-1][thyngd]){
            continue;
        }
        else{
            thyngd = thyngd - weights[i-1];
            curr = curr - values[i-1];
            count = count + 1;
            //cout << "pusha i-1: i indices:" << i-1 << endl;
            indices.push_back(i-1);
        }
    }
    cout << count << endl;
    for(int i : indices){
        cout << i << " ";
    }
    return 0;
}


int main(){
    int C, n;
    while (cin >> C >> n){
       
        
        vector<int> weights;
        vector<int> values;
        
        int thyngdStak, valueStak;
        int count = 0;
        for(int i=0; i<n; i++){
            cin >> valueStak >> thyngdStak;
            values.push_back(valueStak);
            weights.push_back(thyngdStak);
        }
        knap(C, n, weights, values);
        










    }
}
